Support passing a custom server name parameter on HTTPS connection

multani opened the pull request:
## What do these changes do?

This adds the missing support to set the `server_hostname` setting when creating TCP connection, when the underlying connection is authenticated using TLS.

See the documentation for the 2 stdlib functions:

* https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_connection
* https://docs.python.org/3/library/asyncio-eventloop.html#opening-network-connections


This would be needed to support features in clients using aiohttp, such as https://github.com/tomplus/kubernetes_asyncio/issues/267

## Are there changes in behavior for the user?

The default behavior should not change, but this would allow on a per-connection basis to specify a custom server name to check the certificate name against.

## Related issue number

Closes: https://github.com/aio-libs/aiohttp/issues/7114

(for reference, similar implementation in urllib3: https://github.com/urllib3/urllib3/pull/1397)

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

Dreamsorcerer commented on aiohttp/connector.py:
```diff
@@ -1085,14 +1085,18 @@ async def _start_tls_connection(
         # maintainability-wise but this is to be solved separately.
         sslcontext = cast(ssl.SSLContext, self._get_ssl_context(req))
 
+        server_hostname = req.host
+        if req.server_hostname is not None:
+            server_hostname = req.server_hostname
```
Should just be able to do:
```suggestion
        server_hostname = req.server_hostname or req.host
```
Maybe worth moving it back inline as well.

Dreamsorcerer commented on aiohttp/connector.py:
```diff
@@ -1174,6 +1178,10 @@ def drop_exception(fut: "asyncio.Future[List[Dict[str, Any]]]") -> None:
 
+            server_hostname = hinfo["hostname"] if sslcontext else None
+            if req.server_hostname is not None:
+                server_hostname = req.server_hostname
```
Should this be set regardless of `sslcontext`? Seems like a change in behaviour there.

Dreamsorcerer replied:
> Should only be passed if ssl is not None.

So, looks like not.

multani replied:
Implemented the one-liner suggestion below :+1:

Dreamsorcerer commented on tests/test_connector.py:
```diff
@@ -557,6 +557,40 @@ async def certificate_error(*args, **kwargs):
+async def test_tcp_connector_server_hostname_default(loop) -> None:
+    conn = aiohttp.TCPConnector(loop=loop)
+
+    conn._loop.create_connection = mock.AsyncMock()
+    conn._loop.create_connection.return_value = mock.Mock(), mock.Mock()
```
Would be great if you can get this to work with something like:

```suggestion
    with mock.patch.object(conn._loop, "create_connection", autospec=True, spec_set=True):
```

Dreamsorcerer commented on tests/test_connector.py:
```diff
@@ -557,6 +557,40 @@ async def certificate_error(*args, **kwargs):
+
+    req = ClientRequest("GET", URL("https://127.0.0.1:443"), loop=loop)
+
+    established_connection = await conn.connect(req, [], ClientTimeout())
```
Can probably use `async with` here, rather than needing to close explicitly.

multani replied:
Sadly, `.connect` returns a [`Connection`](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/connector.py#L72) which doesn't implement the context management protocol :/

I can use [`contextlib.closing`](https://docs.python.org/3/library/contextlib.html#contextlib.closing) though :+1:

Dreamsorcerer commented:
Also, if the tests still don't work on 3.8 due to missing AsyncMock, feel free to add a `pytest.skipif` for 3.8.

Dreamsorcerer commented on aiohttp/connector.py:
```diff
@@ -1174,6 +1178,10 @@ def drop_exception(fut: "asyncio.Future[List[Dict[str, Any]]]") -> None:
             host = hinfo["host"]
             port = hinfo["port"]
 
+            server_hostname = hinfo["hostname"] if sslcontext else None
+            if req.server_hostname is not None:
+                server_hostname = req.server_hostname
```
```suggestion
            server_hostname = req.server_hostname or hinfo["hostname"] if sslcontext else None
```

Dreamsorcerer commented on docs/client_reference.rst:
```diff
@@ -487,6 +488,13 @@ The client session supports the context manager protocol for self closing.
 
+      :param str server_hostname: Sets or overrides the hostname that the
+         target server’s certificate will be matched against.
+         See `upstream documentation <https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_connection>`_
```
You can use intersphinx:
```suggestion
         See :method:`asyncio.loop.create_connection`
```

multani commented:
@Dreamsorcerer thanks for all the good comments! :bow: 

I made all the changes requested, it's better now! :+1: I'll keep an eye on the CI.

Dreamsorcerer commented:
Oh, I just realised you're pointing to the wrong branch. You need to target a PR to master.

multani commented:
> Oh, I just realised you're pointing to the wrong branch. You need to target a PR to master.

Arghhh, sorry about that, I did my tests through [another library](https://github.com/tomplus/kubernetes_asyncio/pull/270) that was using 3.8 and I used that branch for my personal PR before publishing it here :facepalm: 

Sorry about the inconvenience, I just rebased/squashed all my changes on top of the master branch and changed the PR to point on that branch instead :bow:

multani commented:
@Dreamsorcerer thanks a lot for the merge!

I can take care of the failing backport and create the related pull request if you want :+1:

Dreamsorcerer commented:
Thanks, that would be great. Just follow the instructions and create a PR to 3.9 branch.

multani commented:
The backport is in https://github.com/aio-libs/aiohttp/pull/7543!
